pymc/tests/test_sampling.py

```diff
@@ -1446,3 +1448,207 @@ def test_no_init_nuts_compound(caplog):
+            nested_mu = pm.Normal("nested_mu", mu, 1, size=10)
+            sigma = pm.HalfNormal("sigma", 1)
+            pm.Normal("obs", nested_mu, sigma, observed=y)
+
+        f = compile_forward_sampling_function(
+            outputs=model.observed_RVs,
+            vars_in_trace=[nested_mu, sigma],  # mu isn't in the trace and will be deemed volatile
+            basic_rvs=model.basic_RVs,
+        )
+        assert {i.name for i in self.get_function_inputs(f)} == {"sigma"}
+        assert {i.name for i in self.get_function_roots(f)} == {"sigma"}
+
+        f = compile_forward_sampling_function(
+            outputs=model.observed_RVs,
+            vars_in_trace=[mu, nested_mu, sigma],
+            basic_rvs=model.basic_RVs,
+            givens_dict={
+                mu: np.array(1.0)
+            },  # mu will be considered volatile because it's in givens
+        )
+        assert {i.name for i in self.get_function_inputs(f)} == {"sigma"}
+        assert {i.name for i in self.get_function_roots(f)} == {"mu", "sigma"}
+
+    def test_distributions_op_from_graph(self):
+        with pm.Model() as model:
+            w = pm.Dirichlet("w", a=np.ones(3), size=(5, 3))
+
+            mu = pm.Normal("mu", mu=np.arange(3), sigma=1)
+
+            components = pm.Normal.dist(mu=mu, sigma=1, size=w.shape)
+            mix_mu = pm.Mixture("mix_mu", w=w, comp_dists=components)
+            obs = pm.Normal("obs", mix_mu, 1, observed=np.ones((5, 3)))
+
+        f = compile_forward_sampling_function(
+            outputs=[obs],
+            vars_in_trace=[mix_mu, mu, w],
+            basic_rvs=model.basic_RVs,
+        )
+        assert {i.name for i in self.get_function_inputs(f)} == {"w", "mu", "mix_mu"}
+        assert {i.name for i in self.get_function_roots(f)} == {"mix_mu"}
+
+        f = compile_forward_sampling_function(
+            outputs=[obs],
+            vars_in_trace=[mu, w],
+            basic_rvs=model.basic_RVs,
+        )
+        assert {i.name for i in self.get_function_inputs(f)} == {"w", "mu"}
+        assert {i.name for i in self.get_function_roots(f)} == {"w", "mu"}
+
+        f = compile_forward_sampling_function(
+            outputs=[obs],
+            vars_in_trace=[mix_mu, mu],
+            basic_rvs=model.basic_RVs,
+        )
+        assert {i.name for i in self.get_function_inputs(f)} == {"mu"}
+        assert {i.name for i in self.get_function_roots(f)} == {"mu"}
+
+    def test_distributions_no_op_from_graph(self):
+        with pm.Model() as model:
+            mu = pm.Censored("mu", pm.Normal.dist(mu=np.arange(3), sigma=1), lower=-1, upper=1)
+
+            obs = pm.Normal("obs", mu, 1, observed=np.ones((10, 3)))
+
+        f = compile_forward_sampling_function(
+            outputs=[obs],
+            vars_in_trace=[mu],
+            basic_rvs=model.basic_RVs,
+        )
+        assert {i.name for i in self.get_function_inputs(f)} == {"mu"}
+        assert {i.name for i in self.get_function_roots(f)} == {"mu"}
```

Comment:
Here is the test that covers the suggested changes above:

```suggestion
        with pm.Model() as model:
            latent_mu = pm.Normal("latent_mu", mu=np.arange(3))
            mu = pm.Censored("mu", pm.Normal.dist(latent_mu, sigma=1), lower=-1, upper=1)
            obs = pm.Normal("obs", mu, 1, observed=np.ones((10, 3)))

        f = compile_forward_sampling_function(
            outputs=[obs],
            vars_in_trace=[latent_mu, mu],
            basic_rvs=model.basic_RVs,
        )
        assert {i.name for i in self.get_function_inputs(f)} == {"latent_mu", "mu"}
        assert {i.name for i in self.get_function_roots(f)} == {"mu"}

        f = compile_forward_sampling_function(
            outputs=[obs],
            vars_in_trace=[mu],
            basic_rvs=model.basic_RVs,
        )
        assert {i.name for i in self.get_function_inputs(f)} == set()
        assert {i.name for i in self.get_function_roots(f)} == set()
```